rust/src/read.rs

```diff
@@ -640,25 +644,7 @@ impl<'a> Iterator for MessageStream<'a> {
                 }
 
                 Record::Message { header, data } => {
-                    // Messages must have a previously-read channel.
-                    let channel = match self.channeler.get(header.channel_id) {
-                        Some(c) => c,
-                        None => {
-                            break Some(Err(McapError::UnknownChannel(
-                                header.sequence,
-                                header.channel_id,
-                            )))
-                        }
-                    };
-
-                    let m = Message {
-                        channel,
-                        sequence: header.sequence,
-                        log_time: header.log_time,
-                        publish_time: header.publish_time,
-                        data: Cow::Owned(data.into_owned()),
-                    };
-                    break Some(Ok(m));
+                    break Some(Ok((header, data)));
```

Comment:
why do we use this `break <result>` pattern instead of just returning the value from this function?

Reply:
It's for the bit at the end:
```rust
        if !matches!(n, Some(Ok(_))) {
            self.done = true;
        }
        n
```

If something goes wrong, we want the iterator to return the error, but then return `None` for all subsequent calls to `next()` instead of repeating over and over. Hence the `done` flag and the
```rust
        if self.done {
            return None;
        }
```
check up top. Alternatives seem strictly worse (Store the error and return _it_ over and over instead of `None`? Try to read some more after encountering an unrecoverable problem with the file?)

I'm all ears if you have an idea for a more elegant way to do it.